components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -284,50 +295,57 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
+                    }
                 }
-                // Get the matching user list for the given claim.
-                String[] matchedUserList = getUserList(tenantId, key, value);
-                // Check for duplicates with the already retrieved users, if the matched users for the claim.
-                if (!ArrayUtils.isEmpty(matchedUserList)) {
-                    // In the first iteration resultedUserList will be empty.
-                    if (ArrayUtils.isNotEmpty(resultedUserList)) {
-                        // Remove the users from already matched list who are not in the recently matched list.
-                        resultedUserList = getCommonUserEntries(resultedUserList, matchedUserList, key, value);
-                        if (ArrayUtils.isEmpty(resultedUserList)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("No user matched for given claims");
-                            }
-                            return StringUtils.EMPTY;
+                if (!expressionConditionList.isEmpty()) {
+                    Condition operationalCondition = expressionConditionList.get(0);
+                    if (expressionConditionList.size() > 1) {
+                        for (int i = 1; i < expressionConditionList.size(); i++) {
+                            operationalCondition = new OperationalCondition(OperationalOperation.AND.toString(),
+                                    operationalCondition, expressionConditionList.get(i));
                         }
-                    } else {
-                        resultedUserList = matchedUserList;
-                    }
-                } else {
-                    if (log.isDebugEnabled()) {
-                        String message = String.format("No users matched for claim : %s for value : %s", key, value);
-                        log.debug(message);
                     }
-                    return StringUtils.EMPTY;
+                    resultedUserList.addAll(carbonUM.getUserListWithID(operationalCondition, domain,
+                            UserCoreConstants.DEFAULT_PROFILE, 2, 1, null, null));
                 }
             }
-        }
-        // Return matched user.
-        if (ArrayUtils.isNotEmpty(resultedUserList) && resultedUserList.length == 1) {
-            return resultedUserList[0];
-        } else {
+            // Return empty when no users are found.
+            if (resultedUserList.isEmpty()) {
+                return StringUtils.EMPTY;
+            } else if (resultedUserList.size() == 1) {
+                // Return matched user.
+                return resultedUserList.get(0).getDomainQualifiedUsername();
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Multiple users matched for given claims set : " +
+                            Arrays.toString(resultedUserList.toArray()));
+                }
+                throw Utils.handleClientException(
```

Comment:
Can't we check this case within the for loop that iterates each user store?
This way the error case will be short circuited

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/7a46067ae7fa2693f21ee7bda490d7c1bd37ef0e